Tribler/community/tunnel/processes/tunnel_subprocess.py

```diff
@@ -0,0 +1,277 @@
+                                                         RPC_RESPONSE_ERR,
+                                                         RPC_CREATE,
+                                                         RPC_NOTIFY,
+                                                         RPC_SYNC,
+                                                         RPC_MONITOR,
+                                                         RPC_CIRCUIT,
+                                                         RPC_CIRDEAD)
+from Tribler.community.tunnel.processes.rpcprocess import RPCProcess
+from Tribler.community.tunnel.processes.subprocess import Subprocess
+from Tribler.Core.Session import Session
+from Tribler.Core.SessionConfig import SessionStartupConfig
+from Tribler.Core.simpledefs import NTFY_STARTED, NTFY_TRIBLER
+from Tribler.dispersy.candidate import Candidate
+
+
+class TunnelSubprocess(RPCProcess, Subprocess):
+
+    """The child process's view of the parent process
+
+    In other words, this controls all of the constructs
+    required by the parent process. Most importantly it
+    manages a TunnelCommunity.
+
+    TODO/Future work in this file:
+     - Forward notificiations
+     - Forward bartercast statistics
+    """
+
+    def __init__(self):
+        """Initialize a new TunnelSubprocess
+
+        :returns: None
+        """
+        super(TunnelSubprocess, self).__init__()
+
+        self.session_started = False
+        self.session = None
+        self.community = None
+
+        self.register_rpc(RPC_CIRDEAD)
+        self.register_rpc(RPC_NOTIFY)
+        self.register_rpc(RPC_SYNC, auto_serialize=False)
+        self.register_rpc(RPC_CIRCUIT, self.on_rpc_circuit)
+        self.register_rpc(RPC_CREATE, self.on_rpc_create)
+        self.register_rpc(RPC_MONITOR, self.on_rpc_monitor_infohashes)
+
+    @inlineCallbacks
+    def sync(self, data):
+        """Callback for when any SyncDict wants to sync data
+
+        :param data: the data to synchronize
+        :type data: str
+        :returns: None
+        """
+        yield self.send_rpc(RPC_SYNC, data)
+
+    @inlineCallbacks
+    def circuit_dead(self, circuit_id):
+        """Callback for when a circuit is dead
+
+        :param circuit_id: the dead circuit id
+        :type circuit_id: long
+        :returns: None
+        """
+        yield self.send_rpc(RPC_CIRDEAD, (circuit_id,))
+
+    def on_session_started(self, subject, changetype, objectID, *args):
+        """Callback for when the local Session has started
+
+        Executed on the callback thread: we can't do anything here.
+
+        :param subject: the subject
+        :type subject: str
+        :param changetype: the change type
+        :type changetype: str
+        :param objectID: the change object
+        :type objectID: object
+        :param args: the optional arguments
+        :type args: [object]
+        :returns: None
+        """
+        reactor.callFromThread(self.setup_community)
```

Comment:
This notification is now called from the reactor thread in a very recent commit by me.